rimage/transform/parallel_projection.go

```diff
@@ -95,5 +97,214 @@ func (pp *ParallelProjection) PointCloudToRGBD(cloud pointcloud.PointCloud) (*ri
 
 // ImagePointTo3DPoint takes the 2D pixel point and assumes that it represents the X,Y coordinate in mm as well.
 func (pp *ParallelProjection) ImagePointTo3DPoint(pt image.Point, d rimage.Depth) (r3.Vector, error) {
-	return r3.Vector{float64(pt.X), float64(pt.Y), float64(d)}, nil
+	return r3.Vector{X: float64(pt.X), Y: float64(pt.Y), Z: float64(d)}, nil
+}
+
+// ParallelProjectionOntoXZWithRobotMarker allows the creation of a 2D projection of a pointcloud and robot
+// position onto the XZ plane.
+type ParallelProjectionOntoXZWithRobotMarker struct {
+	robotPose *spatialmath.Pose
+}
+
+const (
+	sigmaLevel        = 7    // level of precision for stdev calculation (determined through experimentation)
+	imageHeight       = 1080 // image height
+	imageWidth        = 1080 // image width
+	missThreshold     = 0.49 // probability limit below which the associated point is assumed to be free space
+	hitThreshold      = 0.55 // probability limit above which the associated point is assumed to indicate an obstacle
+	pointRadius       = 1    // radius of pointcloud point
+	robotMarkerRadius = 5    // radius of robot marker point
+)
+
+// PointCloudToRGBD creates an image of a pointcloud in the XZ plane, scaling the points to a standard image
+// size. It will also add a red marker to the map to represent the location of the robot. The returned depthMap
+// is unused and so will always be nil.
+func (ppRM *ParallelProjectionOntoXZWithRobotMarker) PointCloudToRGBD(cloud pointcloud.PointCloud,
+) (*rimage.Image, *rimage.DepthMap, error) {
+	meta := cloud.MetaData()
+
+	if cloud.Size() == 0 {
+		return nil, nil, errors.New("projection point cloud is empty")
+	}
+
+	meanStdevX, meanStdevZ, err := calculatePointCloudMeanAndStdevXZ(cloud)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	maxX := math.Min(meanStdevX.mean+float64(sigmaLevel)*meanStdevZ.stdev, meta.MaxX)
+	minX := math.Max(meanStdevX.mean-float64(sigmaLevel)*meanStdevZ.stdev, meta.MinX)
+	maxZ := math.Min(meanStdevZ.mean+float64(sigmaLevel)*meanStdevZ.stdev, meta.MaxZ)
+	minZ := math.Max(meanStdevZ.mean-float64(sigmaLevel)*meanStdevZ.stdev, meta.MinZ)
+
+	// Change the max and min values to ensure the robot marker can be represented in the output image
+	var robotMarker spatialmath.Pose
+	if ppRM.robotPose != nil {
+		robotMarker = *ppRM.robotPose
+		maxX = math.Max(maxX, robotMarker.Point().X)
+		minX = math.Min(minX, robotMarker.Point().X)
+		maxZ = math.Max(maxZ, robotMarker.Point().Z)
+		minZ = math.Min(minZ, robotMarker.Point().Z)
+	}
+
+	// Calculate the scale factors
+	scaleFactor := calculateScaleFactor(maxX-minX, maxZ-minZ)
```

Comment:
If the scale factor is 0 this will return an image which is black right? Is that ever useful to an end user or would it be better to return an error in this case?